File: examples/python/cube.py
#!/usr/bin/env python3
"""Parallel to examples/lua/cube.lua: generate Gaussian-cube files
(electron density + electrostatic potential) from a Hartree–Fock
wavefunction. Pair with the isosurface example to render surfaces
colored by ESP in your favorite viewer.

Run:
    python examples/python/cube.py [molecule.xyz]
"""

import sys
import occpy


def main():
    xyz_path = sys.argv[1] if len(sys.argv) > 1 else "examples/scf/water.xyz"
    n = 60

    print(f"Loading: {xyz_path}")
    mol = occpy.Molecule.from_xyz_file(xyz_path)

    print("Running HF/STO-3G ...")
    basis = occpy.AOBasis.load(mol.atoms(), "sto-3g")
    hf = occpy.HartreeFock(basis)
    scf = hf.scf()
    scf.set_charge_multiplicity(0, 1)
    energy = scf.run()
    print(f"SCF energy = {energy:.10f} Ha")

    wfn = scf.wavefunction()

    print(f"\nGenerating {n}x{n}x{n} cubes ...")

    density_cube = occpy.generate_electron_density_cube(wfn, n, n, n)
    with open("density.cube", "w") as f:
        f.write(density_cube)
    print(f"  density.cube       ({len(density_cube)} bytes)")

    esp_cube = occpy.generate_esp_cube(wfn, n, n, n)
    with open("esp.cube", "w") as f:
        f.write(esp_cube)
    print(f"  esp.cube           ({len(esp_cube)} bytes)")

    # Full control via VolumeCalculator + VolumeGenerationParameters.
    calc = occpy.VolumeCalculator()
    calc.set_wavefunction(wfn)

    params = occpy.VolumeGenerationParameters()
    params.property = occpy.VolumePropertyKind.ElectronDensity
    params.adaptive_bounds = True

    vol = calc.compute_volume(params)
    print(f"\nVolumeData: {vol.nx()}x{vol.ny()}x{vol.nz()} = {vol.total_points()} total points")

    cube = calc.volume_as_cube_string(vol)
    with open("density_adaptive.cube", "w") as f:
        f.write(cube)
    print(f"  density_adaptive.cube ({len(cube)} bytes)")


if __name__ == "__main__":
    main()

File: examples/lua/cube.lua
-- examples/lua/cube.lua
-- Generate Gaussian-cube files (electron density + electrostatic
-- potential) from a Hartree–Fock wavefunction. Pair these with the
-- isosurface example to render the surfaces colored by ESP in your
-- favorite viewer.
--
-- Run:
--     occ lua examples/lua/cube.lua [molecule.xyz]

local xyz_path = arg[1] or "examples/scf/water.xyz"
local n = 60   -- grid points per axis

print("Loading:", xyz_path)
local mol = occ.load_molecule(xyz_path)

print("Running HF/STO-3G ...")
local basis = occ.AOBasis_load(mol:atoms(), "sto-3g")
local hf = occ.HartreeFock(basis)
local scf = hf:scf()
scf:set_charge_multiplicity(0, 1)
local energy = scf:run()
print(string.format("SCF energy = %.10f Ha", energy))

local wfn = scf:wavefunction()

print(string.format("\nGenerating %dx%dx%d cubes ...", n, n, n))

-- Both `generate_*_cube` helpers return the cube file contents as a
-- single string — write it to disk yourself.
local density_cube = occ.generate_electron_density_cube(wfn, n, n, n)
do
    local f = io.open("density.cube", "w")
    f:write(density_cube)
    f:close()
    print(string.format("  density.cube       (%d bytes)", #density_cube))
end

local esp_cube = occ.generate_esp_cube(wfn, n, n, n)
do
    local f = io.open("esp.cube", "w")
    f:write(esp_cube)
    f:close()
    print(string.format("  esp.cube           (%d bytes)", #esp_cube))
end

-- For full control over the volume (different property, grid bounds,
-- spin component, etc.), use VolumeCalculator + VolumeGenerationParameters
-- directly:
local calc = occ.VolumeCalculator()
calc:set_wavefunction(wfn)

local params = occ.VolumeGenerationParameters()
params.property = occ.VolumePropertyKind.ElectronDensity
params.adaptive_bounds = true   -- auto-size around atoms

local vol = calc:compute_volume(params)
print(string.format("\nVolumeData: %dx%dx%d = %d total points",
    vol.nx, vol.ny, vol.nz, vol.total_points))

do
    local cube = calc:volume_as_cube_string(vol)
    local f = io.open("density_adaptive.cube", "w")
    f:write(cube)
    f:close()
    print(string.format("  density_adaptive.cube (%d bytes)", #cube))
end
